In [75]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from sklearn import datasets
from sklearn.decomposition import PCA
from sklearn import preprocessing

sns.set() # set seaborn's default settings

First and Foremost, the dataframe has 5 basically empty rows at the top as downloaded from the source: https://www.phase-trans.msm.cam.ac.uk/map/data/materials/austenitic.data.html   

(There is a naming convention for humans viewing the excel file.)  

This can be fixed by using the skiprows parameter in the read_excel function.

In [76]:
df = pd.read_excel('STMECH_AUS_SS.xls',skiprows=5) # skipping the first 5 rows of the excel file (6th row taken as column names)
df

Cr     Ni    Mo    Mn    Si    Nb    Ti  Zr  Ta    V  ...  \
0     18.7  10.69  0.47  1.56  0.62  0.01  0.04   0   0  0.0  ...   
1     18.7  10.69  0.47  1.56  0.62  0.01  0.04   0   0  0.0  ...   
2     18.7  10.69  0.47  1.56  0.62  0.01  0.04   0   0  0.0  ...   
3     18.7  10.69  0.47  1.56  0.62  0.01  0.04   0   0  0.0  ...   
4     18.7  10.69  0.47  1.56  0.62  0.01  0.04   0   0  0.0  ...   
...    ...    ...   ...   ...   ...   ...   ...  ..  ..  ...  ...   
2175  17.7  12.28  0.00  1.24  0.30  0.83  0.00   0   0  0.0  ...   
2176  17.7  12.28  0.00  1.24  0.30  0.83  0.00   0   0  0.0  ...   
2177  17.7  12.28  0.00  1.24  0.30  0.83  0.00   0   0  0.0  ...   
2178  17.7  12.28  0.00  1.24  0.30  0.83  0.00   0   0  0.0  ...   
2179  17.7  12.28  0.00  1.24  0.30  0.83  0.00   0   0  0.0  ...   

      Grains mm-2  Type of melting  Size of ingot  Product form  \
0      477.712892                1              3             1   
1      477.712892                1              3             1   
2      477.712892                1              3             1   
3      477.712892                1              3             1   
4      477.712892                1              3             1   
...           ...              ...            ...           ...   
2175           Na                2             Na             1   
2176           Na                2             Na             1   
2177           Na                2             Na             1   
2178           Na                2             Na             1   
2179           Na                2             Na             1   

      Temperature (K)  0.2%proof_stress (M Pa)  UTS (M Pa)  Elongation (%)  \
0                 298               280.000000  610.000000              67   
1                 373               243.000000  508.000000              54   
2                 473               220.000000  472.000000              44   
3                 573               204.000000  470.000000              43   
4                 673               197.000000  470.000000              45   
...               ...                      ...         ...             ...   
2175              723               171.616375  454.047895            34.9   
2176              773               166.713050  435.415260              33   
2177              823               169.655045  424.627945            34.2   
2178              873               168.674380  398.149990            37.6   
2179              973               139.254430  296.160830            45.4   

      Area_reduction (%)                Comments  
0                     83  NRIM/CDS/No4B/1986 ABA  
1                     83  NRIM/CDS/No4B/1986 ABA  
2                     81  NRIM/CDS/No4B/1986 ABA  
3                     76  NRIM/CDS/No4B/1986 ABA  
4                     76  NRIM/CDS/No4B/1986 ABA  
...                  ...                     ...  
2175                  Na                     NaN  
2176                  Na                     NaN  
2177                  Na                     NaN  
2178                  Na                     NaN  
2179                  Na                     NaN  

[2180 rows x 35 columns]

In [77]:
df.index

RangeIndex(start=0, stop=2180, step=1)

In [78]:
df.columns

Index(['Cr', 'Ni', 'Mo', 'Mn', 'Si', 'Nb', 'Ti', 'Zr', 'Ta', 'V', 'W', 'Cu',
       'N', 'C', 'B', 'P', 'S', 'Co', 'Al', 'Sn', 'Pb',
       'Solution_treatment_temperature', 'Solution_treatment_time(s)',
       'Water_Quenched_after_s.t.', 'Air_Quenched_after_s.t.', 'Grains mm-2',
       'Type of melting', 'Size of ingot', 'Product form', 'Temperature (K)',
       '0.2%proof_stress (M Pa)', 'UTS (M Pa)', 'Elongation (%)',
       'Area_reduction (%)', 'Comments'],
      dtype='object')

In [79]:
# Lets rename the columns with easier names for our convenience
df = df.rename(columns={'Solution_treatment_temperature':'sol_temp',
                        'Solution_treatment_time(s)': 'sol_time',
                        'Water_Quenched_after_s.t.':'water_quench',
                        'Air_Quenched_after_s.t.':'air_quench',
                        'Grains mm-2':'grains',
                        'Type of melting':'melting_type',
                        'Size of ingot':'ingot_size',
                        'Product form':'product_form',
                        'Temperature (K)': 'temp',
                        '0.2%proof_stress (M Pa)': 'proof_stress',
                        'UTS (M Pa)': 'uts',
                        'Elongation (%)': 'elongation',
                        'Area_reduction (%)': 'area_reduction',
                        'Comments': 'comments'})
df.head()

Cr     Ni    Mo    Mn    Si    Nb    Ti  Zr  Ta    V  ...      grains  \
0  18.7  10.69  0.47  1.56  0.62  0.01  0.04   0   0  0.0  ...  477.712892   
1  18.7  10.69  0.47  1.56  0.62  0.01  0.04   0   0  0.0  ...  477.712892   
2  18.7  10.69  0.47  1.56  0.62  0.01  0.04   0   0  0.0  ...  477.712892   
3  18.7  10.69  0.47  1.56  0.62  0.01  0.04   0   0  0.0  ...  477.712892   
4  18.7  10.69  0.47  1.56  0.62  0.01  0.04   0   0  0.0  ...  477.712892   

   melting_type  ingot_size  product_form  temp  proof_stress    uts  \
0             1           3             1   298         280.0  610.0   
1             1           3             1   373         243.0  508.0   
2             1           3             1   473         220.0  472.0   
3             1           3             1   573         204.0  470.0   
4             1           3             1   673         197.0  470.0   

   elongation  area_reduction                comments  
0          67              83  NRIM/CDS/No4B/1986 ABA  
1          54              83  NRIM/CDS/No4B/1986 ABA  
2          44              81  NRIM/CDS/No4B/1986 ABA  
3          43              76  NRIM/CDS/No4B/1986 ABA  
4          45              76  NRIM/CDS/No4B/1986 ABA  

[5 rows x 35 columns]

In [80]:
df.shape

(2180, 35)

In [81]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2180 entries, 0 to 2179
Data columns (total 35 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Cr              2180 non-null   float64
 1   Ni              2180 non-null   float64
 2   Mo              2180 non-null   float64
 3   Mn              2180 non-null   float64
 4   Si              2180 non-null   float64
 5   Nb              2180 non-null   float64
 6   Ti              2180 non-null   float64
 7   Zr              2180 non-null   int64  
 8   Ta              2180 non-null   int64  
 9   V               2180 non-null   float64
 10  W               2179 non-null   float64
 11  Cu              2180 non-null   float64
 12  N               2180 non-null   float64
 13  C               2180 non-null   float64
 14  B               2180 non-null   float64
 15  P               2180 non-null   float64
 16  S               2180 non-null   float64
 17  Co              2180 non-null   f

We can organize our data to get two main distinct lists for categorical and numerical columns.  

In [82]:
cats = list(df.select_dtypes(include=['object']).columns)
nums = list(df.select_dtypes(exclude=['object']).columns)
print(f'categorical variables:  {cats}')
print(f'numerical variables:  {nums}')

categorical variables:  ['sol_temp', 'sol_time', 'water_quench', 'air_quench', 'grains', 'melting_type', 'ingot_size', 'elongation', 'area_reduction', 'comments']
numerical variables:  ['Cr', 'Ni', 'Mo', 'Mn', 'Si', 'Nb', 'Ti', 'Zr', 'Ta', 'V', 'W', 'Cu', 'N', 'C', 'B', 'P', 'S', 'Co', 'Al', 'Sn', 'Pb', 'product_form', 'temp', 'proof_stress', 'uts']


We can also inspect how many unique values are there in each column. This is useful to identify categorical variables that may have been misclassified as numerical due to their data type.

In [83]:
df.nunique(axis=0) 

Cr                132
Ni                145
Mo                 79
Mn                 71
Si                 54
Nb                 30
Ti                 54
Zr                  1
Ta                  1
V                  12
W                   1
Cu                 27
N                  48
C                  29
B                  27
P                  32
S                  32
Co                 18
Al                 41
Sn                  1
Pb                  1
sol_temp           20
sol_time           20
water_quench        3
air_quench          3
grains             43
melting_type        5
ingot_size         21
product_form        5
temp               20
proof_stress      371
uts               597
elongation        298
area_reduction     66
comments           60
dtype: int64

We can see from the output of our above nunquie() funciton that 5 is a value that can and should be considered categorical in this case espeically with how many rows of data we have.  
Beyond 5, the next number of uniques jumps up to 20+ for non physical element columns. This is a good indicator that we should consider 5 uniques as a categorical variable.

We can then see what those 5 or less unique values are with the following:

In [84]:
for col in df.columns:
    if df[col].nunique() <= 5:
        print(f'{col}: {df[col].unique()}')

Zr: [0]
Ta: [0]
W: [ 0. nan]
Sn: [0]
Pb: [0]
water_quench: [1 'Na' 0]
air_quench: [0 'Na' 1]
melting_type: [1 2 3 'Na' 4]
product_form: [1 2 3 4 5]


Ignoring the physical elements, we can see the that those are indeed categorical variables.

water_quench, air_quench, and melting_type, are already set as categorical in our 'cats' list, but product_form is in our numerical list, so lets fix that:

In [85]:
cats.extend(['product_form'])
nums.remove('product_form')

print(f'categorical variables:  {cats}')
print(f'numerical variables:  {nums}')

categorical variables:  ['sol_temp', 'sol_time', 'water_quench', 'air_quench', 'grains', 'melting_type', 'ingot_size', 'elongation', 'area_reduction', 'comments', 'product_form']
numerical variables:  ['Cr', 'Ni', 'Mo', 'Mn', 'Si', 'Nb', 'Ti', 'Zr', 'Ta', 'V', 'W', 'Cu', 'N', 'C', 'B', 'P', 'S', 'Co', 'Al', 'Sn', 'Pb', 'temp', 'proof_stress', 'uts']


## Inspect for `nan`s in data.

At this point we have noticed that some 'Na' values are appearing. From our df.info() funciton we saw non-null values for all columns. 

This is because the 'Na' values are strings and not NaN values. We can fix this by replacing all 'Na' values with NaN values, and then doing some null value inspeciton and dropping rows that include them if it is a small enough part of our dataset.

In [86]:
df = df.replace('Na', np.nan)
df.tail() # saw from earlier in the notebook the last 5 rows have 'Na' values so we can make sure they were replaced with NaN

Cr     Ni   Mo    Mn   Si    Nb   Ti  Zr  Ta    V  ...  grains  \
2175  17.7  12.28  0.0  1.24  0.3  0.83  0.0   0   0  0.0  ...     NaN   
2176  17.7  12.28  0.0  1.24  0.3  0.83  0.0   0   0  0.0  ...     NaN   
2177  17.7  12.28  0.0  1.24  0.3  0.83  0.0   0   0  0.0  ...     NaN   
2178  17.7  12.28  0.0  1.24  0.3  0.83  0.0   0   0  0.0  ...     NaN   
2179  17.7  12.28  0.0  1.24  0.3  0.83  0.0   0   0  0.0  ...     NaN   

      melting_type  ingot_size  product_form  temp  proof_stress         uts  \
2175           2.0         NaN             1   723    171.616375  454.047895   
2176           2.0         NaN             1   773    166.713050  435.415260   
2177           2.0         NaN             1   823    169.655045  424.627945   
2178           2.0         NaN             1   873    168.674380  398.149990   
2179           2.0         NaN             1   973    139.254430  296.160830   

      elongation  area_reduction  comments  
2175        34.9             NaN       NaN  
2176        33.0             NaN       NaN  
2177        34.2             NaN       NaN  
2178        37.6             NaN       NaN  
2179        45.4             NaN       NaN  

[5 rows x 35 columns]

In [87]:
df.isnull().any()

Cr                False
Ni                False
Mo                False
Mn                False
Si                False
Nb                False
Ti                False
Zr                False
Ta                False
V                 False
W                  True
Cu                False
N                 False
C                 False
B                 False
P                 False
S                 False
Co                False
Al                False
Sn                False
Pb                False
sol_temp           True
sol_time           True
water_quench       True
air_quench         True
grains             True
melting_type       True
ingot_size         True
product_form      False
temp              False
proof_stress      False
uts               False
elongation         True
area_reduction     True
comments           True
dtype: bool

In [88]:
df.isna().sum()

Cr                   0
Ni                   0
Mo                   0
Mn                   0
Si                   0
Nb                   0
Ti                   0
Zr                   0
Ta                   0
V                    0
W                    1
Cu                   0
N                    0
C                    0
B                    0
P                    0
S                    0
Co                   0
Al                   0
Sn                   0
Pb                   0
sol_temp           184
sol_time          1097
water_quench       264
air_quench         264
grains            1517
melting_type       217
ingot_size        1517
product_form         0
temp                 0
proof_stress         0
uts                  0
elongation          95
area_reduction    1517
comments          1517
dtype: int64

In [89]:
df.isna().sum().sum()

8190

In [90]:
# nan proportion in data
8190 / len(df)

3.756880733944954

We have too many rows to drop all rows with NaN values, so we will drop rows with NaN values in the following columns:

In [91]:
# Checking if null values are from the same rows for large null value columns
columns_with_many_nulls = df.isna().sum() > 200
selected_columns = df.columns[columns_with_many_nulls]
df_nulls = df[selected_columns]
df_nulls.isna().sum()

sol_time          1097
water_quench       264
air_quench         264
grains            1517
melting_type       217
ingot_size        1517
area_reduction    1517
comments          1517
dtype: int64

In [92]:
null_rows = df_nulls.isnull().all(axis=1)
null_rows.sum()

3

This shows only 3 rows there there are full NaN rows. We can drop those for sure, but there are way too many other NaN values that should be dropped.  
Lets do that again but increase the `columns_with_many_nulls` parameter to look for > 300, not just 200:

In [93]:
# Checking if null values are from the same rows for large null value columns
columns_with_many_nulls = df.isna().sum() > 300
selected_columns = df.columns[columns_with_many_nulls]
df_nulls = df[selected_columns]
df_nulls.isna().sum()

sol_time          1097
grains            1517
ingot_size        1517
area_reduction    1517
comments          1517
dtype: int64

In [94]:
null_rows = df_nulls.isnull().all(axis=1)
null_rows.sum()

800

We can see now that we have 800 rows that are fully NaN on these columns. We can drop those rows:

In [95]:
df.shape

(2180, 35)

In [98]:
df_firstdrop = df.loc[~null_rows]
df_firstdrop.shape

(1380, 35)

We can see from our two df.shape command that those 800 all null value rows (for selected columns) have been dropped from the dataframe.

At this point, we could do other methods to only drop rows if the `NaN` value is in certian columns, but for the sake of this case as we are not heading into creating a model yet, just prep, we will just drop all columns with NaN values. (We can come back to our df_firstdrop dataframe if we want to do more analysis on/around the NaN values later.)

In [101]:
df = df_firstdrop.dropna(axis=0)
df.shape

(365, 35)

In [103]:
365/2180

0.16743119266055045

So, we are only keeping 16.7% of the inital data set. This is a good amount of data to work with, but we should keep in mind that we are only working with 16.7% of the original data set.

In [108]:
print(f'totalduplicaterows:{df.duplicated().sum()}') 

totalduplicaterows:0
